src/index.ts

```diff
@@ -0,0 +1,84 @@
+import express, {Request, Response} from 'express';
+import {v4} from 'uuid';
+import User from './user-actions';
+import * as userActions from './user-actions';
+import * as userValidation from './validation';
+import {sha512} from 'js-sha512';
+
+const app = express();
+const port = 3000;
+
+const users: User[] = [];
+
+app.use(express.json());
+
+app.get('/user', (req: Request, res: Response) => {
+  const validationResult = userValidation.userSchema.query.validate(req.query);
+  if (validationResult.error) {
+    res.status(400).send();
+  } else {
+    const loginSubstring = req.query.loginSubstring || '';
+    const limit = req.query.limit ? +req.query.limit : undefined;
+
+    res.send(
+      userActions.getResUsers(
+        userActions.getAutoSuggestUsers(users, loginSubstring.toString(), limit)
+      )
+    );
+  }
+});
+
+app.get('/user/:id', (req: Request, res: Response) => {
+  const user = users.find(user => user.id === req.params.id);
+  if (user && !user.isDeleted) {
+    res.send(userActions.getResUser(user));
+  } else {
+    res.status(404).send();
+  }
+});
+
+app.post('/user', (req: Request, res: Response) => {
+  if (userValidation.validateUserParams(req, res)) {
+    const user: User = {
+      id: v4(),
+      login: req.body.login,
+      password: sha512(req.body.password),
+      age: req.body.age,
+      isDeleted: false,
+    };
+    users.push(user);
+    res.send(userActions.getResUser(user));
+  }
+});
+
+app.put('/user', (req: Request, res: Response) => {
+  if (userValidation.validateUserQuery(req, res)) {
+    const userIndex = users.findIndex(user => user.id === req.body.id);
+    if (userIndex >= 0) {
+      if (users[userIndex].isDeleted) {
+        res.status(500).send();
+      } else {
+        users[userIndex].login = req.body.login;
+        users[userIndex].password = req.body.password;
+        users[userIndex].age = req.body.age;
+        res.send(userActions.getResUser(users[userIndex]));
+      }
+    } else {
+      res.status(404).send();
```

Comment:
I'm almost sure you send response twice
* first time in userValidation.validateUserQuery `res.status(400).send();`
* second time - here, when validation function returns false you send response again

I think it's better to refactor the approach because it's feels bad that validation function sends response so you should  keep an eye on several places where `res.send()` can be called.

These points below related to WHOLE validation
* extract validation in separate middleware
* refactor validation function to, for example, return errors
* in validation middleware check if there are any errors and return bad request response or pass execution further
* you can take a look at https://www.npmjs.com/package/express-joi-validation. Looks like it provides middlewares for express

Reply:
and if validation fails you should return detailed error

Reply:
removed those validation functions, no need in them with express-joi-validation
validadtion returns message like this
![image](https://user-images.githubusercontent.com/14372066/213553622-d5b4a922-da86-4519-b17f-f61d495d29c1.png)

Reply:
added global exception for 505 and moved messages into separate file